src/test/groovy/com/synacy/lesson04/exercise/service/BankAccountServiceTest.groovy

```diff
@@ -73,8 +77,112 @@ class BankAccountServiceTest extends Specification {
+		bankAccount.getBalance() >> new BigDecimal(2258.25)
+
+		when:
+		bankAccountService.deposit(bankAccount, 700.00)
+
+		then:
+		1 * bankAccount.setBalance(2958.25)
+
+		then:
+		1 * bankAccountDao.saveBankAccount(bankAccount)
+
+	}
+
+	def "deposit should record the transaction made by the account"() {
+		given:
+		def bankAccount = Mock(BankAccount)
+		bankAccount.getBalance() >> new BigDecimal(2258.25)
+		def amountDeposited = 700.00
+		def date = Mock(Date)
+
+		when:
+		bankAccountService.deposit(bankAccount, amountDeposited)
+
+		then:
+		1 * transactionDao.saveTransaction(*_) >> { Transaction transaction ->
+			assert bankAccount == transaction.bankAccount
+			assert TransactionType.DEBIT == transaction.type
+			assert amountDeposited == transaction.amount
+			assert date == transaction.transactionDate
+		}
+	}
+	def "deposit should throw an exception at deposits less than 200.00"() {
+		given:
+		def bankAccount = Mock(BankAccount)
+		def amountDeposited = 199.99
+
+
+		when:
+		bankAccountService.deposit(bankAccount, amountDeposited)
+
+		then:
+		InsufficientDepositException exception = thrown()
+		bankAccount == exception.bankAccount
+		amountDeposited == exception.amountToDeposit
+	}
+
+	def "fetchAllTransactions should get all transactions for a given bank account"() {
+		given:
+		def bankAccount = Mock(BankAccount)
+		def transaction1 = Mock(Transaction)
+		def transaction2 = Mock(Transaction)
+		def date1 = Mock(Date)
+		def date2 = Mock(Date)
+		transaction1.transactionDate >> date1
+		transaction2.transactionDate >> date2
+
+		def expectedTransactions = [
+		        transaction1, transaction2
+		]
+
+		transactionDao.fetchAllTransactionsOfBankAccount(bankAccount) >> expectedTransactions
+
+		when:
+		def actualTransactions = bankAccountService.fetchAllTransactions(bankAccount)
+
+		then:
+		actualTransactions.containsAll(expectedTransactions)
+	}
+
+	def "fetchAllTransactions should return all transactions from a given bank account by date by order of most recent"() {
+		given:
+		def bankAccount = Mock(BankAccount)
+
+		def transaction1 = Mock(Transaction)
+		def transaction2 = Mock(Transaction)
+		def transaction3 = Mock(Transaction)
+
+		def date1 = Mock(Date)
+		def date2 = Mock(Date)
+		def date3 = Mock(Date)
+
+//		transaction1.transactionDate >> date1
+//		transaction2.transactionDate >> date2
+//		transaction3.transactionDate >> date3
+
+		transaction1.transactionDate >> new SimpleDateFormat("MM-dd-yyyy").parse("05-31-2017")
+		transaction2.transactionDate >> new SimpleDateFormat("MM-dd-yyyy").parse("05-20-2017")
+		transaction3.transactionDate >> new SimpleDateFormat("MM-dd-yyyy").parse("05-04-2017")
+
+		def expectedTransactions = [
+		        transaction1, transaction2, transaction3
+		]
+
+		transactionDao.fetchAllTransactionsOfBankAccount(bankAccount) >> expectedTransactions
+
+		expect:
+		expectedTransactions.asList() == bankAccountService.fetchAllTransactions(bankAccount)
+
+	}
+
+
```

Comment:
No test for transfer money? :crying_cat_face:  Even though the implementation is  a mere reuse of the methods withdraw and deposit, tests should be supplied because of the following reasons:
 
 * TDD requires you to write tests first.
 * This is a little bit of a blackbox testing. Whatever your implementations may be, it must conform to the tests made. So even though a test for both `deposit` and `withdraw` methods exist, that does not mean you don't have to provide test for `transfer` method. The tests for transfer method is independent from the tests of `deposit` and `withdraw` methods.
 * The transfer method may have additional features in the future that isn't the same from both deposit and withdraw methods.